File: rs/earthbucks_lib/tests/iso_buf_reader_test.rs
use byteorder::{BigEndian, WriteBytesExt};
use earthbucks_lib::iso_buf_reader::*;
use serde::Deserialize;
use std::fs;

#[test]
fn test_read() {
    let mut reader = IsoBufReader::new(vec![1, 2, 3, 4, 5]);
    assert_eq!(reader.read(3).unwrap(), vec![1, 2, 3]);
    assert_eq!(reader.read(2).unwrap(), vec![4, 5]);
}

#[test]
fn test_read_u8() {
    let mut reader = IsoBufReader::new(vec![1, 2, 3, 4, 5]);
    assert_eq!(reader.read_u8().unwrap(), 1);
    assert_eq!(reader.read_u8().unwrap(), 2);
}

#[test]
fn test_read_u16_be() {
    let mut buffer_reader = IsoBufReader::new(vec![0x01, 0x23]);
    assert_eq!(buffer_reader.read_u16_be().unwrap(), 0x0123);
}

#[test]
fn test_read_u32_be() {
    let mut data = vec![];
    data.write_u32::<BigEndian>(1234567890).unwrap();
    data.write_u32::<BigEndian>(987654321).unwrap();

    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_u32_be().unwrap(), 1234567890);
    assert_eq!(reader.read_u32_be().unwrap(), 987654321);
}

#[test]
fn test_read_u64_be_big_int() {
    let mut data = vec![];
    data.write_u64::<BigEndian>(12345678901234567890).unwrap();
    data.write_u64::<BigEndian>(9876543210987654321).unwrap();

    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_u64_be().unwrap(), 12345678901234567890);
    assert_eq!(reader.read_u64_be().unwrap(), 9876543210987654321);
}

#[test]
fn test_read_var_int_buf() {
    let data = vec![0xfd, 0x01, 0x00];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_var_int_buf().unwrap(), vec![0xfd, 0x01, 0x00]);

    let data = vec![0xfe, 0x01, 0x00, 0x00, 0x00];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(
        reader.read_var_int_buf().unwrap(),
        vec![0xfe, 0x01, 0x00, 0x00, 0x00]
    );

    let data = vec![0xff, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(
        reader.read_var_int_buf().unwrap(),
        vec![0xff, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00]
    );

    let data = vec![0x01];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_var_int_buf().unwrap(), vec![0x01]);
}

#[test]
fn test_read_var_int() {
    let data = vec![0xfd, 0x10, 0x01];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_var_int().unwrap(), 0x1000 + 1);

    let data = vec![0xfe, 0x10, 0x00, 0x00, 0x01];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_var_int().unwrap(), 0x10000000 + 1);

    let data = vec![0xff, 0x10, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_var_int().unwrap(), 0x1000000000000000 + 1);

    let data = vec![0x01];
    let mut reader = IsoBufReader::new(data);
    assert_eq!(reader.read_var_int().unwrap(), 1);
}

// standard test vectors

#[derive(Deserialize)]
struct TestVectorIsoBufReader {
    read: TestVectorReadIsoBuf,
    read_u8: TestVectorReadErrors,
    read_u16_be: TestVectorReadErrors,
    read_u32_be: TestVectorReadErrors,
    read_u64_be: TestVectorReadErrors,
    read_var_int_buf: TestVectorReadErrors,
    read_var_int: TestVectorReadErrors,
}

#[derive(Deserialize)]
struct TestVectorReadIsoBuf {
    errors: Vec<TestVectorReadIsoBufError>,
}

#[derive(Deserialize)]
struct TestVectorReadIsoBufError {
    hex: String,
    len: usize,
    error: String,
}

#[derive(Deserialize)]
struct TestVectorReadErrors {
    errors: Vec<TestVectorReadError>,
}

#[derive(Deserialize)]
struct TestVectorReadError {
    hex: String,
    error: String,
}

#[test]
fn test_vectors_read() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read(test_vector.len);
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert!(e.to_string().starts_with(&test_vector.error)),
        }
    }
}

#[test]
fn test_vectors_read_u8() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read_u8.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read_u8();
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert_eq!(e.to_string(), test_vector.error),
        }
    }
}

#[test]
fn test_vectors_read_u16_be() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read_u16_be.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read_u16_be();
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert!(e.to_string().starts_with(&test_vector.error)),
        }
    }
}

#[test]
fn test_vectors_read_u32_be() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read_u32_be.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read_u32_be();
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert!(e.to_string().starts_with(&test_vector.error)),
        }
    }
}

#[test]
fn test_vectors_read_u64_be() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read_u64_be.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read_u64_be();
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert!(e.to_string().starts_with(&test_vector.error)),
        }
    }
}

#[test]
fn test_vectors_read_var_int_buf() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read_var_int_buf.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read_var_int_buf();
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert!(e.to_string().starts_with(&test_vector.error)),
        }
    }
}

#[test]
fn test_vectors_read_var_int() {
    let data =
        fs::read_to_string("./test_vectors/iso_buf_reader.json").expect("Unable to read file");
    let test_vectors: TestVectorIsoBufReader =
        serde_json::from_str(&data).expect("Unable to parse JSON");
    for test_vector in test_vectors.read_var_int.errors {
        let buf = hex::decode(&test_vector.hex).expect("Failed to decode hex");
        let mut reader = IsoBufReader::new(buf);
        let result = reader.read_var_int();
        match result {
            Ok(_) => panic!("Expected an error, but got Ok(_)"),
            Err(e) => assert!(e.to_string().starts_with(&test_vector.error)),
        }
    }
}

File: ts/earthbucks-iso-buf/test/iso-buf-reader.test.ts
import { describe, expect, test, beforeEach } from "vitest";
import { Reader } from "../src/reader.js";
import { Buffer } from "buffer";
import { IsoBuf } from "../src/iso-buf.js";
import fs from "fs";
import path from "path";

describe("Reader", () => {
  let isoBufReader: Reader;
  let testBuffer: IsoBuf;

  beforeEach(() => {
    testBuffer = new IsoBuf(new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8]));
    isoBufReader = new Reader(testBuffer);
  });

  test("read returns correct subarray", () => {
    const len = 4;
    const result = isoBufReader.read(len).unwrap();
    expect(result).toEqual(testBuffer.subarray(0, len));
  });

  test("read updates position", () => {
    const len = 4;
    isoBufReader.read(len);
    expect(isoBufReader["pos"]).toBe(len);
  });

  test("readU8 returns correct value and updates position", () => {
    const result = isoBufReader.readU8().unwrap();
    expect(result).toBe(1);
    expect(isoBufReader["pos"]).toBe(1);
  });

  test("readU16BE returns correct value and updates position", () => {
    const result = isoBufReader.readU16BE().unwrap();
    expect(result).toBe(IsoBuf.from([1, 2]).unwrap().readU16BE(0).unwrap());
    expect(isoBufReader["pos"]).toBe(2);
  });

  test("readU32BE returns correct value and updates position", () => {
    const result = isoBufReader.readU32BE().unwrap();
    expect(result).toBe(
      IsoBuf.from([1, 2, 3, 4]).unwrap().readU32BE(0).unwrap(),
    );
    expect(isoBufReader["pos"]).toBe(4);
  });

  test("readU64BEBigInt returns correct value and updates position", () => {
    // Create a BufferReader with a buffer that contains the 64-bit unsigned integer 0x0123456789ABCDEF
    isoBufReader = new Reader(
      new IsoBuf(
        new Uint8Array([0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef]),
      ),
    );

    const result = isoBufReader.readU64BE().unwrap();

    // Check that the method returns the correct BigInt
    expect(result).toEqual(BigInt("0x0123456789ABCDEF"));

    // Check that the position has been updated correctly
    expect(isoBufReader["pos"]).toBe(8);
  });

  test("readVarIntNum returns correct value and updates position for small numbers", () => {
    const result = isoBufReader.readVarIntNum().unwrap();
    expect(result).toBe(1); // Assuming that the implementation treats a single byte as a varint
    expect(isoBufReader["pos"]).toBe(1);
  });

  test("readVarIntNum returns correct value and updates position for 16 bit numbers", () => {
    const buf = IsoBuf.from([0xfd, 0, 0, 0, 0]).unwrap();
    buf.writeU16BE(500, 1);
    isoBufReader = new Reader(buf); // A varint that represents the number 2^30
    const result = isoBufReader.readVarIntNum().unwrap();
    expect(result).toBe(500); // 2^30
    expect(isoBufReader["pos"]).toBe(3);
  });

  test("readVarIntNum returns correct value and updates position for 32 bit numbers", () => {
    const buf = IsoBuf.from([254, 0, 0, 0, 0]).unwrap();
    buf.writeU32BE(2000000000, 1);
    isoBufReader = new Reader(buf); // A varint that represents the number 2^30
    const result = isoBufReader.readVarIntNum().unwrap();
    expect(result).toBe(2000000000); // 2^30
    expect(isoBufReader["pos"]).toBe(5);
  });

  test("readVarIntNum", () => {
    let bufferReader = new Reader(IsoBuf.from([0xfd, 0x00, 0x01]).unwrap());
    expect(bufferReader.readVarIntNum().val.toString()).toBe(
      "non-minimal encoding",
    );

    bufferReader = new Reader(
      IsoBuf.from([0xfe, 0x00, 0x00, 0x00, 0x01]).unwrap(),
    );
    expect(bufferReader.readVarIntNum().val.toString()).toBe(
      "non-minimal encoding",
    );

    bufferReader = new Reader(
      IsoBuf.from([
        0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
      ]).unwrap(),
    );
    expect(bufferReader.readVarIntNum().val.toString()).toBe(
      "non-minimal encoding",
    );

    bufferReader = new Reader(IsoBuf.from([0x01]).unwrap());
    expect(bufferReader.readVarIntNum().unwrap()).toBe(1);
  });

  test("readVarIntBuf", () => {
    let bufferReader = new Reader(IsoBuf.from([0xfd, 0x00, 0x01]).unwrap());
    expect(bufferReader.readVarIntBuf().val.toString()).toEqual(
      "non-minimal encoding",
    );

    bufferReader = new Reader(
      IsoBuf.from([0xfe, 0x00, 0x00, 0x00, 0x01]).unwrap(),
    );
    expect(bufferReader.readVarIntBuf().val.toString()).toEqual(
      "non-minimal encoding",
    );

    bufferReader = new Reader(
      IsoBuf.from([
        0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
      ]).unwrap(),
    );
    expect(bufferReader.readVarIntBuf().val.toString()).toEqual(
      "non-minimal encoding",
    );

    bufferReader = new Reader(IsoBuf.from([0x01]).unwrap());
    expect(bufferReader.readVarIntBuf().unwrap()).toEqual(
      IsoBuf.from([0x01]).unwrap(),
    );
  });

  test("readVarInt", () => {
    let bufferReader = new Reader(IsoBuf.from([0xfd, 0x00, 0x01]).unwrap());
    expect(bufferReader.readVarInt().val.toString()).toEqual(
      "non-minimal encoding",
    );

    bufferReader = new Reader(
      IsoBuf.from([0xfe, 0x00, 0x00, 0x00, 0x01]).unwrap(),
    );
    expect(bufferReader.readVarInt().val.toString()).toEqual(
      "non-minimal encoding",
    );

    bufferReader = new Reader(
      IsoBuf.from([
        0xff, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x01,
      ]).unwrap(),
    );
    expect(bufferReader.readVarInt().val.toString()).toEqual(
      "non-minimal encoding",
    );

    bufferReader = new Reader(IsoBuf.from([0x01]).unwrap());
    expect(bufferReader.readVarInt().unwrap()).toEqual(BigInt(1));
  });

  describe("test vectors", () => {
    interface TestVectorReader {
      read: TestVectorReadIsoBuf;
      read_u8: TestVectorReadErrors;
      read_u16_be: TestVectorReadErrors;
      read_u32_be: TestVectorReadErrors;
      read_u64_be: TestVectorReadErrors;
      read_var_int_buf: TestVectorReadErrors;
      read_var_int: TestVectorReadErrors;
    }

    interface TestVectorReadIsoBuf {
      errors: TestVectorReadIsoBufError[];
    }

    interface TestVectorReadIsoBufError {
      hex: string;
      len: number;
      error: string;
    }

    interface TestVectorReadErrors {
      errors: TestVectorReadError[];
    }

    interface TestVectorReadError {
      hex: string;
      error: string;
    }

    const filePath = path.resolve(__dirname, "../test-vectors/reader.json");
    const jsonString = fs.readFileSync(filePath, "utf-8");
    const testVector: TestVectorReader = JSON.parse(jsonString);

    test("test vectors: read", () => {
      testVector.read.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(bufferReader.read(test.len).val.toString()).toEqual(test.error);
      });
    });

    test("test vectors: read_u8", () => {
      testVector.read_u8.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(
          bufferReader.readU8().val.toString().startsWith(test.error),
        ).toBeTruthy();
      });
    });

    test("test vectors: read_u16_be", () => {
      testVector.read_u16_be.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(
          bufferReader.readU16BE().val.toString().startsWith(test.error),
        ).toBeTruthy();
      });
    });

    test("test vectors: read_u32_be", () => {
      testVector.read_u32_be.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(
          bufferReader.readU32BE().val.toString().startsWith(test.error),
        ).toBeTruthy();
      });
    });

    test("test vectors: read_u64_be", () => {
      testVector.read_u64_be.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(
          bufferReader.readU64BE().val.toString().startsWith(test.error),
        ).toBeTruthy();
      });
    });

    test("test vectors: read_var_int_buf", () => {
      testVector.read_var_int_buf.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(bufferReader.readVarIntBuf().val.toString()).toMatch(
          new RegExp("^" + test.error.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")),
        );
      });
    });

    test("test vectors: read_var_int", () => {
      testVector.read_var_int.errors.forEach((test) => {
        const buf = IsoBuf.from(test.hex, "hex").unwrap();
        const bufferReader = new Reader(buf);
        expect(bufferReader.readVarInt().val.toString()).toMatch(
          new RegExp("^" + test.error.replace(/[.*+?^${}()|[\]\\]/g, "\\$&")),
        );
      });
    });
  });
});
